### PR TITLE
cli: git uploadrefs

### PR DESCRIPTION
**Pull request by samelhusseini**
Add cli support to directly upload git refs to the cloud. 

CLI command recurses through all git references (commit, tree, blob) and checks the cloud for their status in chunks (50 per request). It then uploads to the cloud any missing hashes. 

Finally it calls the cloud to update the git refs config file.

### CODE REVIEW

File: cli/gitfs.ts [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+
+export type SMap<T> = { [s: string]: T };
+
+interface QEntry {
+    run: () => Promise<any>;
+    resolve: (v: any) => void;
+    reject: (err: any) => void;
+}
+
+export function promiseQueue() {
```
Comment (mmoskal) [inline cli/gitfs.ts]: Better use Util.PromiseQueue

File: cli/gitfs.ts [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        this.cache[key] = v
+    }
+
+    flush() {
+        this.size = 0
+        this.cache = {}
+    }
+}
+
+export class PromiseBuffer<T> {
```
Comment (mmoskal) [inline cli/gitfs.ts]: Better use Util.PromiseBuffer

File: cli/gitfs.ts (hunk 1 above)
Comment (samelhusseini) [reply cli/gitfs.ts]: makes sense

File: cli/gitfs.ts (hunk 2 above)
Comment (samelhusseini) [reply cli/gitfs.ts]: makes sense

**Comment by samelhusseini** [comment]
Ready.

**Pull request merged**